Modix/Modix.csproj

```diff
@@ -1,33 +1,98 @@
     <PackageReference Include="Serilog.Sinks.RollingFile" Version="3.3.0" />
     <PackageReference Include="Serilog.Sinks.Sentry" Version="2.1.6" />
   </ItemGroup>
+
   <ItemGroup>
-    <ProjectReference Include="..\Modix.Services\Modix.Services.csproj" />
-    <ProjectReference Include="..\Modix.Data\Modix.Data.csproj" />
-    <ProjectReference Include="..\Modix.WebServer\Modix.WebServer.csproj" />
+    <ClientSrc Include="$(SpaRoot)**" Exclude="$(SpaRoot)node_modules\**" />
   </ItemGroup>
+
   <ItemGroup>
-    <None Update=".dockerignore">
-      <DependentUpon>Dockerfile</DependentUpon>
-    </None>
+    <!-- Don't publish the SPA source files, but do show them in the project files list -->
+    <Content Remove="$(SpaRoot)**" />
+    <None Include="$(SpaRoot)**" Exclude="$(SpaRoot)node_modules\**" />
   </ItemGroup>
+
   <ItemGroup>
-    <Folder Include="Properties\" />
+    <ProjectReference Include="..\Modix.Bot\Modix.Bot.csproj" />
+    <ProjectReference Include="..\Modix.Data\Modix.Data.csproj" />
+    <ProjectReference Include="..\Modix.Services\Modix.Services.csproj" />
   </ItemGroup>
-</Project>
\ No newline at end of file
+
+  <Target Name="EnsureNodeEnv" BeforeTargets="Build" Condition=" '$(Configuration)' == 'Debug' And !Exists('$(SpaRoot)node_modules') ">
+    <!-- Ensure Node.js is installed -->
+    <Exec Command="node --version" ContinueOnError="true">
+      <Output TaskParameter="ExitCode" PropertyName="ErrorCode" />
+    </Exec>
+    <Error Condition="'$(ErrorCode)' != '0'" Text="Node.js is required to build and run this project. To continue, please install Node.js from https://nodejs.org/, and then restart your command prompt or IDE." />
+  </Target>
+
+  <Target Name="NpmInstall" DependsOnTargets="EnsureNodeEnv" Inputs="$(SpaRoot)package-lock.json" Outputs="$(SpaRoot)node_modules\.npm_timestamp">
+    <Message Importance="high" Text="Restoring dependencies using 'npm'. This may take several minutes..." />
+    <Exec WorkingDirectory="$(SpaRoot)" Command="npm install" />
+
+    <!-- Touch an empty file so we know the last time we ran restore. If the timestamp on this file is greater than the
+         timestamp on the package-lock.json file then this build step will be skipped. -->
+    <Touch Files="$(SpaRoot)node_modules\.npm_timestamp" AlwaysCreate="true">
+      <Output TaskParameter="TouchedFiles" ItemName="FileWrites" />
+    </Touch>
+  </Target>
+
+  <Target Name="RunVueCli" BeforeTargets="Build" DependsOnTargets="NpmInstall" Inputs="@(ClientSrc)" Outputs="$(MSBuildThisFileDirectory)obj\$(Configuration)\.vue_timestamp">
+    <Exec WorkingDirectory="$(SpaRoot)" Command="npm run $(NpmBuild)" />
+
+    <Touch Files="$(MSBuildThisFileDirectory)obj\$(Configuration)\.vue_timestamp" AlwaysCreate="true">
+      <Output TaskParameter="TouchedFiles" ItemName="FileWrites" />
+    </Touch>
+  </Target>
+
+  <Target Name="VueFirstBuild" BeforeTargets="RunVueCli" Condition=" !Exists('wwwroot\index.html') ">
```

Comment:
I would like to see a build configuration that doesn't also build the frontend - for cases of quick iteration on the backend - but it's not a huge priority as it seems to no-op decently fast enough when there's no changes. I tried to investigate this myself in conjunction with launchSettings.json, but got nowhere.